[TRAINING] Ruby Exercises

Deathshoo7 opened the pull request:
The purpose of the pr is add exercises about the training feature in monokera.

Deathshoo7 committed:
initial version

ajalvarez89 commented on adivinar_numero.rb:
```diff
@@ -0,0 +1,11 @@
+n = rand(1..10)
+puts "Bienvenido al programa, Adivina el número que estoy pensando"
```
ruby convention is `'` for string without interpolation and `"` for interpolations

ajalvarez89 commented on adivinar_numero.rb:
```diff
@@ -0,0 +1,11 @@
+puts "Bienvenido al programa, Adivina el número que estoy pensando"
+puts "Por favor ingresa el número"
+num = gets.chomp.to_i
+while num!=n
```
```suggestion
while num != n
```

ajalvarez89 commented on adivinar_numero.rb:
```diff
@@ -0,0 +1,11 @@
+while num!=n
+    puts "Intenta de nuevo. "
+    puts "Por favor ingresa el número"
+    num = gets.chomp.to_i
```
for this particular case the variable would be valid but keep in mind that it is a good practice to use telling names.
```suggestion
    number = gets.chomp.to_i
```

juannicolasvargas replied:
ruby es un lenguaje muy diciente, por lo que importa mucho que el nombre de cada variable o clase o servicios sea congruente a lo que va  a hacer o almacenar, en este caso como almacena un numero, el nombre de la variable number queda mejor que num, de tal modo que si otro dev la ve sabe que guarda

ajalvarez89 commented on adivinar_numero.rb:
```diff
@@ -0,0 +1,11 @@
+    num = gets.chomp.to_i
+end
+
+puts "Felicitaciones, lo lograste! el numero era:#{num}"
```
```suggestion
puts "Felicitaciones, lo lograste! el numero era: #{ num }"
```

ajalvarez89 commented on adivinar_numero.rb:
```diff
@@ -0,0 +1,11 @@
+n = rand(1..10)
```
```suggestion
number_random = rand(1..10)
```

ajalvarez89 commented on bmi.rb:
```diff
@@ -0,0 +1,7 @@
+puts "Por favor ingresa tu altura"
+height = gets.chomp.to_f
+imcRound =  (weight / (height**2)).round
+puts "Tu BMI es de: #{imcRound}"
```
```suggestion
puts "Tu BMI es de: #{ imcRound }"

```

ajalvarez89 commented on bmi.rb:
```diff
@@ -0,0 +1,7 @@
+weight = gets.chomp.to_i
+puts "Por favor ingresa tu altura"
+height = gets.chomp.to_f
+imcRound =  (weight / (height**2)).round
```
in ruby ​​the convention for variables and methods is snake_case
```suggestion
imc_round =  (weight / (height ** 2)).round
```

ajalvarez89 commented on letas_repetidas.rb:
```diff
@@ -0,0 +1,41 @@
+    end
+end
+
+def getDuplicates(firstArr, secondArr, repetidasArr)
```
```suggestion
def get_duplicates(first_array, second_array, repeted_array)
```

ajalvarez89 commented on print_list.rb:
```diff
@@ -0,0 +1,27 @@
+puts "Bienvenido al programa, Imprimir lista"
+puts "Ingresa el numero de personas"
+lenghtList = gets.chomp.to_i
+personas = Array.new
```
check this [ruby-style-guide#literal-array-hash](https://github.com/rubocop/ruby-style-guide#literal-array-hash)
```suggestion
personas = []
```

ajalvarez89 commented:
The general indentation must be two spaces.

ajalvarez89 commented on edad.rb:
```diff
@@ -0,0 +1,13 @@
+msg = user_age>current_year ? "Imposible! Aún no has nacido" 
+                                : current_age == 1 ? "Tienes 1 año" : "Tienes #{current_age} años" 
+        
+
+
```
✂️ avoid extra lines.

ajalvarez89 commented on edad.rb:
```diff
@@ -0,0 +1,13 @@
+puts "Por favor ingresa el año en que naciste."
+user_age = gets.chomp.to_i
+current_age = current_year-user_age
+msg = user_age>current_year ? "Imposible! Aún no has nacido" 
```
```suggestion
msg = user_age > current_year ? 'Imposible! Aún no has nacido'
```

ajalvarez89 commented on edad.rb:
```diff
@@ -0,0 +1,13 @@
+puts "Bienvenido al calculador de edad."
+puts "Por favor ingresa el año en que naciste."
+user_age = gets.chomp.to_i
+current_age = current_year-user_age
```
```suggestion
current_age = current_year_user_age
```

juannicolasvargas commented on edad.rb:
```diff
@@ -0,0 +1,13 @@
+require 'date'
```
It was super good, it is not an error but as a general rule we always try to do everything in English in ruby, names of variables, methods, models, classes at least it is the standard we have here, so as not to combine English and Spanish as it usually happens, for what the age class should go in English, so that you keep it in mind.

ajalvarez89 commented on edad.rb:
```diff
@@ -0,0 +1,13 @@
+require 'date'
```
is not required, we could use `Time.now.year`

ajalvarez89 replied:
✂️

ajalvarez89 commented on edad.rb:
```diff
@@ -0,0 +1,13 @@
+require 'date'
+# Obtenemos el año actual.
+current_year = DateTime.now.year
```
```suggestion
current_year = Time.now.year
```

ajalvarez89 commented on edad.rb:
```diff
@@ -0,0 +1,13 @@
+        
+
+
+puts msg
```
```suggestion
puts msg

```

ajalvarez89 commented on letas_repetidas.rb:
```diff
@@ -0,0 +1,41 @@
+    firstArrWords = firstWord.chars
+    secondArrWords = secondWord.chars
+    repetidasArr = []
+    if firstArrWords.length <= secondArrWords.length
+        getDuplicates(firstArrWords, secondArrWords, repetidasArr)
+    else
+        getDuplicates(secondArrWords, firstArrWords, repetidasArr)        
+    end    
+
+    if(repetidasArr.length > 0)
+        
+        repetidasArrSort = repetidasArr.sort
+        letrasRepetidas =""
+        repetidasArrSort.each do |v|
+            letrasRepetidas += v +", "
+        end        
+        puts "Letras: #{letrasRepetidas}"
+        puts "Total letras: #{repetidasArrSort.length}"
+    end
+end
+
+def getDuplicates(firstArr, secondArr, repetidasArr)
+    firstArr.each do |v|
+        if !v !="" && secondArr.include?(v)
+            if repetidasArr.select{|w| w == v}.length == 0
+                repetidasArr.push(v)
+            end
+            
+        end
+    end
+end
+
+
+puts "Bienvenido al programa, Detectar letras repetidas"
+puts "Ingresa la primera palabra."
+firstWord = gets.chomp
+puts "Ingresa la segunda palabra."
+secondWord = gets.chomp
+puts repetidas(firstWord, secondWord)
+
```
check this enhancement:
```suggestion
def words_repeted(words)
    repeted_words = get_duplicated_words(words.first.chars, words.last.chars)

  unless (repeted_words.empty?)
    puts "Letras: #{ repeted_words.sort.join(', ') }"
    puts "Total letras: #{ repeted_words.length }"
  end
end

def get_duplicated_words(first_word, second_word)
  first_word.select { |letter| second_word.include?(letter) }.uniq!
end

puts 'Bienvenido al programa, Detectar letras repetidas'
puts 'Ingresa dos palabras'
words = gets.chomp.split(' ')
puts words_repeted(words)
```

ajalvarez89 commented on traductor.rb:
```diff
@@ -0,0 +1,17 @@
+def traducir(palabra)
+    tblTraductor ={"uno"=>"one", "dos"=>"two","tres"=>"three","cuatro"=>"four","cinco"=>"five","seis"=>"six","siete"=>"seven","ocho"=>"eight","nueve"=>"nine","diez"=>"ten",}        
+    number_lowercase = palabra.downcase
+    bNumberExist = tblTraductor.has_key?(number_lowercase)
+    if(!bNumberExist)
+        puts "No encontramos la traducción de: #{number_lowercase}"
+    else
+        puts "La traducción de: #{number_lowercase} es: #{tblTraductor[number_lowercase]}"
+
+    end
+end
+
+
+puts "Bienvenido al programa, Traductor de números"
+puts "Ingresa numero a traducir"
+number_to_translate = gets.chomp
+puts traducir(number_to_translate)
```
```suggestion
def translate(word)
  if(dictionary.has_key?(word.to_sym))
    puts "La traducción de: #{ word } es: #{ dictionary[word.to_sym] }"
  else
    puts "No encontramos la traducción de: #{ word }"
  end
end

def dictionary
  {
    uno: 'one',
    dos: 'two',
    tres: 'three',
    cuatro: 'four',
    cinco: 'five',
    seis: 'six',
    siete: 'seven',
    ocho: 'eight',
    nueve: 'nine',
    diez: 'ten'
  }.freeze
end


puts 'Bienvenido al programa, Traductor de números'
puts 'Ingresa numero a traducir'
number_to_translate = gets.chomp
puts translate(number_to_translate.downcase)
```

ajalvarez89 commented on suma.rb:
```diff
@@ -0,0 +1,8 @@
+puts "Bienvenido al programa para sumar 2 números."
+puts "Por favor ingresa el primer número"
+firstNumber = gets.chomp.to_i
```
```suggestion
first_number = gets.chomp.to_i
```

ajalvarez89 commented on suma.rb:
```diff
@@ -0,0 +1,8 @@
+puts "Por favor ingresa el primer número"
+firstNumber = gets.chomp.to_i
+puts "Por favor ingresa el segundo número"
+secondNumber = gets.chomp.to_i
```
```suggestion
second_number = gets.chomp.to_i
```

ajalvarez89 commented on suma.rb:
```diff
@@ -0,0 +1,8 @@
+firstNumber = gets.chomp.to_i
+puts "Por favor ingresa el segundo número"
+secondNumber = gets.chomp.to_i
+puts "Los números ingresados son: #{firstNumber} y #{secondNumber}"
```
```suggestion
puts "Los números ingresados son: #{ first_number } y #{ second_number }"
```

ajalvarez89 commented on suma.rb:
```diff
@@ -0,0 +1,8 @@
+secondNumber = gets.chomp.to_i
+puts "Los números ingresados son: #{firstNumber} y #{secondNumber}"
+#total = firstNumber + secondNumber;
+puts "La suma total de números ingresados es: #{firstNumber + secondNumber}"
```
```suggestion
puts "La suma total de números ingresados es: #{ first_number + second_number }"

```

ajalvarez89 commented on saluda.rb:
```diff
@@ -0,0 +1,4 @@
+#Preguntar nombre
+puts "Hola, por favor ingresa tu nombre para saludarte."
+username = gets.chomp
+puts "Hola #{username}, bienvenido !!"
```
```suggestion
puts "Hola #{ username }, bienvenido !!"

```

ajalvarez89 commented on repite.rb:
```diff
@@ -0,0 +1,7 @@
+#puts "Hello world Fuck Yeaaaa"
+# Interpolacion/
+#puts "Interpolacion del año #{2021}"}
+puts "Escribe una frase:"
```
```suggestion
puts 'Escribe una frase:'
```

ajalvarez89 commented on repite.rb:
```diff
@@ -0,0 +1,7 @@
+#puts "Interpolacion del año #{2021}"}
+puts "Escribe una frase:"
+name = gets.chomp
+puts "La frase ingresada es:#{name}"
```
```suggestion
puts "La frase ingresada es: #{ name }"
```

ajalvarez89 commented on menor_que_diez.rb:
```diff
@@ -0,0 +1,5 @@
+puts "Bienvenido al programa. El numero es Mayor que diez?."
+puts "Por favor ingresa un número"
+n = gets.chomp.to_i
+msg = n<=10 ? "El número es menor o igual a 10" : "El número es mayor a 10"
+puts msg
```
```suggestion
puts 'Bienvenido al programa. El numero es Mayor que diez?.'
puts 'Por favor ingresa un número'
number = gets.chomp.to_i
message = number <= 10 ? 'El número es menor o igual a 10' : 'El número es mayor a 10'
puts message
```

Deathshoo7 committed:
modificado:     adivinar_numero.rb
	modificado:     bmi.rb
	modificado:     edad.rb
	modificado:     letas_repetidas.rb
	modificado:     menor_que_diez.rb
	modificado:     print_list.rb
	modificado:     repite.rb
	modificado:     saluda.rb
	modificado:     suma.rb
	modificado:     traductor.rb
